command/agent/retry_join_test.go

```diff
@@ -170,6 +171,44 @@ func TestRetryJoin_Server_MixedProvider(t *testing.T) {
 	require.Equal(stubAddress, output[0])
 }
 
+func TestRetryJoin_AutoDiscover(t *testing.T) {
+	ci.Parallel(t)
+	require := require.New(t)
+
+	var output []string
+
+	mockJoin := func(s []string) (int, error) {
+		output = s
+		return 0, nil
+	}
+
+	// 'exec=*'' test go-netaddr
+	// 'localhost' also tetss get-netaddr by ussuring it resolves to "::1" and "127.0.0.1"
+	// '100.100.100.100' ensures that bare IPs are used as-is
+	serverJoin := &ServerJoin{
+		RetryMaxAttempts: 1,
+		RetryJoin:        []string{"exec=echo 192.168.1.1 192.168.1.2", "localhost", "100.100.100.100", "provider=aws, tag_value=foo"},
+	}
+
+	joiner := retryJoiner{
+		discover:      discoverProxy{goDiscover: &discover.Discover{}},
+		serverJoin:    mockJoin,
+		serverEnabled: true,
+		logger:        testlog.HCLogger(t),
+		errCh:         make(chan struct{}),
+	}
+
+	joiner.RetryJoin(serverJoin)
+
+	require.Equal(6, len(output))
+	require.Equal("192.168.1.1", output[0])
+	require.Equal("192.168.1.2", output[1])
+	require.Equal("::1", output[2])
+	require.Equal("127.0.0.1", output[3])
```

Comment:
These assertions on `::1` and `127.0.0.1` are system-dependent and will likely need to be refactored. In order to ensure that `go-netaddrs` resolves hostnames to IP addresses, a stable DNS entry must be in place. Currently, this relies on `/etc/hosts` to have entries for `::1` and `127.0.0.1` corresponding with the name `localhost`.

Reply:
The best approach is to not try to test the `go-netaddrs` and `go-discovery` libraries here by testing only the logic that's in Nomad. That might require a little bit of refactoring to make work.